File: 02/RomanArabic.cpp
#include "RomanArabic.hpp"

#include <stdexcept>
#include <utility>
#include <string>
#include <array>

std::string RomanArabic::Arabic2Roman(int number)
{
    if (number <= 0 || number > 3999)
        throw Exception{std::to_string(number) + " is out of range"};

    std::string result;
    for (const auto& [roman, value] : numbers)
    {
        if (number - value >= 0)
        {
            result += roman;
            number -= value;
        }
    }

    if (number != 0)
        throw "Something very wrong happened";

    return result;
}

int RomanArabic::Roman2Arabic(const std::string &str)
{
    int sum = 0;
    int current_index = 0;
    bool was_compound = false; // for combinations of symbols with one of lower value before one of higher e.g. IX, IV

    for (int i = 0; i < numbers.size() / 3; i++)
    {
        bool row_used = false; // a row is 3 elements, see definition of numbers (if it is still formatted right)
        for (int j = 0; j < 3; j++)
        {
            const auto& [roman, value] = numbers[3 * i + j];
            if (str.find(roman, current_index) == current_index)
            {
                if (row_used || was_compound)
                    throw Exception{"The roman numerals " + str + " have a wrong format"};
                row_used = true;
                was_compound = i % 2 == 1 && j != 1;
                current_index += roman.length();
                sum += value;
            }
        }
        was_compound = was_compound && i % 2 == 1;
    }

    if (current_index != str.size())
        throw Exception{"The roman numerals " + str + " have a wrong format"};

    return sum;
}

const std::array<std::pair<std::string, int>, 21> RomanArabic::numbers
    {{{"MMM", 3000}, {"MM", 2000}, { "M", 1000},
      { "CM",  900}, { "D",  500}, {"CD",  400},
      {"CCC",  300}, {"CC",  200}, { "C",  100},
      { "XC",   90}, { "L",   50}, {"XL",   40},
      {"XXX",   30}, {"XX",   20}, { "X",   10},
      { "IX",    9}, { "V",    5}, {"IV",    4},
      {"III",    3}, {"II",    2}, { "I",    1}}};

File: 02/RomanArabic.java
public class RomanArabic {

    public static int Roman2Arabic(String str) throws RomanArabicException {
        int sum = 0;
        int currentIndex = 0;
        boolean wasCompound = false; // for combinations of symbols with one of lower value before one of higher e.g. IX, IV

        for (int i = 0; i < allRomanNumerals.length / 3; i++) {
            boolean rowUsed = false; // a row is 3 elements
            for (int j = 0; j < 3; j++) {
                if (str.indexOf(allRomanNumerals[3 * i + j], currentIndex) == currentIndex) {
                    if (rowUsed || wasCompound)
                        throw new RomanArabicException("The roman numerals " + str + " have a wrong format");
                    rowUsed = true;
                    wasCompound = i % 2 == 1 && j != 1;
                    currentIndex += allRomanNumerals[3 * i + j].length();
                    sum += allValues[3 * i + j];
                }
            }
            wasCompound = wasCompound && i % 2 == 1;
        }

        if (currentIndex != str.length())
            throw new RomanArabicException("The roman numerals " + str + " have a wrong format");

        return sum;
    }

    public static String Arabic2Roman(int number) throws RomanArabicException {
        if (number <= 0 || number > 3999)
            throw new RomanArabicException(number + " is out of range");

        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < allRomanNumerals.length; i++) {
            if (number - allValues[i] >= 0) {
                builder.append(allRomanNumerals[i]);
                number -= allValues[i];
            }
        }

        if (number != 0)
            throw new RuntimeException("Something very wrong happened");

        return builder.toString();
    }

    private static final String[] allRomanNumerals =
        { "MMM", "MM",  "M",
           "CM",  "D", "CD",
          "CCC", "CC",  "C",
           "XC",  "L", "XL",
          "XXX", "XX",  "X",
           "IX",  "V", "IV",
          "III", "II",  "I"};

    private static final int[] allValues = 
        {  3000, 2000, 1000,
            900,  500,  400,
            300,  200,  100,
             90,   50,   40,
             30,   20,   10,
              9,    5,    4,
              3,    2,    1};

}